docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -110,6 +111,77 @@ We need to be able to process transactions and roll-back state updates if a tran
+
+The latter indicates, however, that the implementation doesn't provide true modularity. Instead it causes problems related to race condition and sync (see: [\#6370](https://github.com/cosmos/cosmos-sdk/issues/6370)).
+
+We propose to reduce the multistore concept from the SDK, and to use a single instance of `SC` and `SS` in a `rootStore` object. To avoid confusion, we should rename the `MultiStore` interface to `RootStore`.
+
+Moreover, to improve usability, we should extend the `KVStore` interface with _prefix store_. This will allow module developers to bind a store to a namespace for module sub-components:
+
+```go
+type KVStore interface {
+    ... current KVStore
+
+    WithPrefix(prefix string) KVStore
+}
+```
+
+The `WithPrefix` method will create a proxy object for the parent object and will prepend `prefix` to a key parameter of all key-value operations (similar to the `store/prefix.Store` implementation).
+The following invariant must hold:
+
+```
+for each OP in [Get Has, Set, ...]
+    store.WithPrefix(prefix).OP(key) == store.OP(prefix + key)
+```
+
+
+The `RootStore` will have the following interface:
+
+```
+TODO
+```
+
+In contrast to `MultiStore`, `RootStore` doesn't allow to mount dynamically sub stores. However it can inherit a multistore functionality in it's concrete implementation (which will be needed for IBC support).
+
+#### Merkle Proofs and IBC
+
+Currently IBC (v1.0) module merkle proof for a `(key, value)` consists of two elements `[storeKey, proof]`. Verification using 2 passes:
+1. Checks that a `proof` is a valid merkle proof `(key, value)` using ICS-23 spec.
+2. Then it checks that the `storeKey` and `root(proof)` hashes to the AppHash (App state commitment).
+
+Breaking this behavior would severely impact the Cosmos ecosystem which already widely adopts the IBC module. Unfortunately, the straightforward implementation is breaking because all keys in `SC` are hashed. So we can't simply make the last step. Even if we set the `storeKey` to an empty string it will rehash.
+
+For workaround we need to:
++ keep the double hashing and multistore concept for IBC.
++ the `RootStore` will have only two stores: the general one, and another for IBC. `RootStore` should not expose mounting stores in a "runtime" (it's only possible to do it through constructor).
++ The App Hash is a hash of both stores in the `RootStore`.
+
+
+
+### Optimization: compress module keys
+
+We consider a compression of prefix keys using [Huffman Coding](https://en.wikipedia.org/wiki/Huffman_coding). It will require knowledge of used prefixes (module store keys) a priori. And for best results, it will need frequency information for each prefix (how often objects are stored in the store under the same prefix key). With Huffman Coding, the `WithPrefix` invariant has the following form:
+
+```
+for each OP in [Get Has, Set, ...]
+    store.WithPrefix(prefix).OP(key) == store.OP(store.Code(prefix) + key)
```

Comment:
Right, `store.Set(...)` doesn't know if it should uses a compression or not.
So, to avoid confusions we should enable the compression only on the "first" level (when doing `ctx.KVStore`). I will update the paragraph.